### ISSUE TITLE
Infinite recursion on Crystal master when running on its own code

### ISSUE THREAD

**Issue by refi64** [elbywan/crystalline#24]
This is certainly an interesting bug...

Basically, given:

- A Crystal compiler built from `master`
- Crystalline running on its own source code

Crystalline will segfault due to infinite recursion. Here's what I've gathered so far:

- A bisect seems to point to [this PR](https://github.com/crystal-lang/crystal/pull/10527) being the start of the breakage.
- The infinite recursion occurs around when the compiler is analyzing `src/compiler/crystal/syntax/parser.cr`, line 4484, where and `MainVisitor.type_assign` is being called on the same node infinitely.

**Comment by elbywan** [comment]
Hi @refi64,

> This is certainly an interesting bug...

Thank you for reporting and investigating!

This is an interesting issue indeed 🤔. I tried to reproduce using the compiler binaries from the 1.1 release, but with my local setup it seems to be working fine.

Maybe this is related to a specific setup, or maybe the bug has been fixed since in the master crystal branch (or something else).

 I'll keep the issue open in any case.
